src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1079,7 +1095,13 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
       .withType(dummy.nonMemberTermRef)
     checkVariance(impl1)
     if (!cls.is(AbstractOrTrait) && !ctx.isAfterTyper) checkRealizableBounds(cls.typeRef, cdef.pos)
-    assignType(cpy.TypeDef(cdef)(name, impl1, Nil), cls)
+    val cdef1 = assignType(cpy.TypeDef(cdef)(name, impl1, Nil), cls)
+    if (ctx.phase.isTyper && cdef1.tpe <:< defn.DynamicType && !ctx.dynamicsEnabled) {
+      val isRequired = parents1.exists(_.tpe =:= defn.DynamicType)
```

Comment:
I reversed engineered that from the warning in scalac of all the test cases I have. As far as I can tell the conditions in scalac are equivalent but checked in a different way. I might also have taken some input from @sjrd of how they are expected to behave from the use cases in Scala.js.